driver-core/src/main/com/mongodb/internal/connection/CommandMessage.java

```diff
@@ -107,30 +111,49 @@ public final class CommandMessage extends RequestMessage {
         this.serverApi = serverApi;
     }
 
+    // Create a BsonDocument representing the logical document encoded by an OP_MSG.
+    // The returned document will contain all the fields, including ones represented by
+    // OP_MSG Sections of type Kind 1: Document Sequence
     BsonDocument getCommandDocument(final ByteBufferBsonOutput bsonOutput) {
-        ByteBufBsonDocument byteBufBsonDocument = ByteBufBsonDocument.createOne(bsonOutput,
-                getEncodingMetadata().getFirstDocumentPosition());
-        BsonDocument commandBsonDocument;
-
-        if (containsPayload()) {
-            commandBsonDocument = byteBufBsonDocument.toBaseBsonDocument();
-
-            int payloadStartPosition = getEncodingMetadata().getFirstDocumentPosition()
-                    + byteBufBsonDocument.getSizeInBytes()
-                    + 1 // payload type
-                    + 4 // payload size
-                    + payload.getPayloadName().getBytes(StandardCharsets.UTF_8).length + 1;  // null-terminated UTF-8 payload name
-            commandBsonDocument.append(payload.getPayloadName(),
-                    new BsonArray(ByteBufBsonDocument.createList(bsonOutput, payloadStartPosition)));
-        } else {
-            commandBsonDocument = byteBufBsonDocument;
+        List<ByteBuf> duplicateByteBuffers = bsonOutput.getByteBuffers();
```

Comment:
The refactoring in ByteBufBsonDocument allows this method to create the list of byte buffers once and re-use them both for creating the BsonDocument from the Kind 0 Section and the List<BsonDocument> for each Kind 1 Sections.